Add Keccak256 inline, ANDN op to tracer

danielwlz opened the pull request:
- Keccak256 (SHA-3) inline added to tracer
- Helper Keccak256 added to jolt-sdk (similar interface to sha3 crate)
- `Keccak256SequenceBuilder` builds the virtual sequence from primitive instructions. Using `VirtualROTRI` and  `ANDN` reduces sequence length to ≈3.3k RV64 instructions (approximately 25 cycles per byte)
- Introduced `ANDN` (AND-with-NOT) instruction with test coverage;
- Migrated `sha3-ex` and `sha3-chain` examples to use the inline
- Added `CpuTestHarness` to streamline unit tests for instructions and inlines
- Fixed VirtualROTRI for RvV64 (Omid)

### Test plan
```bash
# unit + algorithm tests (passing)
cargo test --package tracer --lib -- instruction::inline_keccak256

# sdk integration (passing)
cargo test --package jolt-sdk --features host,keccak256 --lib -- keccak256::tests
```

Tests verify against known input/outputs and verify trace and exec paths match.

### Next steps 
- Add lookup table and prefix-suffix decompositions for `ANDN` to jolt-core
- (Daniel) Move shared emitters from `Keccak256SequenceBuilder` into `inline_helpers` to benefit SHA-2 and future inlines
- (Daniel) Amend SHA2 inline to use added code and instructions


**Request:** please squash this PR when merging.

danielwlz commented:
On local machine:

```
❯ cargo test --package jolt-sdk --features host,keccak256 --lib -- keccak256::tests

warning: Patch `ark-bn254 v0.5.0-alpha.0 (https://github.com/a16z/arkworks-algebra?branch=dev%2Ftwist-shout#efc56e0d)` was not used in the crate graph.
Check that the patched package version and available features are compatible
with the dependency requirements. If the patch has a different version from
what is locked in the Cargo.lock file, run `cargo update` to use the new
version. This may also occur with an optional dependency that is not enabled.
   Compiling tracer v0.2.0 (/Users/dwong/Projects/jolt-sumcheck/jolt/tracer)
   Compiling jolt-core v0.1.0 (/Users/dwong/Projects/jolt-sumcheck/jolt/jolt-core)
   Compiling jolt-sdk v0.1.0 (/Users/dwong/Projects/jolt-sumcheck/jolt/jolt-sdk)
    Finished `test` profile [optimized + debuginfo] target(s) in 3.29s
     Running unittests src/lib.rs (target/debug/deps/jolt_sdk-3ed151f340a01760)

running 1 test
test keccak256::tests::test_keccak256_empty ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

❯ cargo test --package tracer --lib -- instruction::inline_keccak256

warning: Patch `ark-bn254 v0.5.0-alpha.0 (https://github.com/a16z/arkworks-algebra?branch=dev%2Ftwist-shout#efc56e0d)` was not used in the crate graph.
Check that the patched package version and available features are compatible
with the dependency requirements. If the patch has a different version from
what is locked in the Cargo.lock file, run `cargo update` to use the new
version. This may also occur with an optional dependency that is not enabled.
   Compiling tracer v0.2.0 (/Users/dwong/Projects/jolt-sumcheck/jolt/tracer)
    Finished `test` profile [optimized + debuginfo] target(s) in 1.17s
     Running unittests src/lib.rs (target/debug/deps/tracer-47eae98cd197884e)

running 13 tests
test instruction::inline_keccak256::tests::test_execute_chi ... ok
test instruction::inline_keccak256::keccak256::tests::measure_keccak_length ... ok
test instruction::inline_keccak256::keccak256::tests::test_keccak256_direct_execution ... ok
test instruction::inline_keccak256::tests::test_execute_iota ... ok
test instruction::inline_keccak256::tests::test_execute_keccak256 ... ok
test instruction::inline_keccak256::tests::test_execute_keccak_f ... ok
test instruction::inline_keccak256::tests::test_execute_rho_and_pi ... ok
test instruction::inline_keccak256::tests::test_execute_theta ... ok
test instruction::inline_keccak256::tests::test_rotl64 ... ok
test instruction::inline_keccak256::tests::test_step_by_step_round_1 ... ok
test instruction::inline_keccak256::keccak256::tests::test_keccak_against_reference ... ok
test instruction::inline_keccak256::keccak256::tests::test_keccak_exec_trace_equal ... ok
test instruction::inline_keccak256::keccak256::tests::test_keccak_exec_trace_intermediate_vr_equal ... ok

test result: ok. 13 passed; 0 failed; 0 ignored; 0 measured; 24 filtered out; finished in 0.07s

```

moodlezoup reviewed (commented):
great work! Just a couple of nits

danielwlz commented on tracer/src/instruction/inline_keccak256/test_utils.rs:
```diff
@@ -0,0 +1,254 @@
+
+    /// Load a Keccak state into DRAM and set `x10 = BASE_ADDR`.
+    pub fn load_state(&mut self, state: &Keccak256State) {
+        self.harness.cpu.x[10] = Self::BASE_ADDR as i64;
```
Good catch! I thought I made that fix but must have regressed.

danielwlz commented:
@moodlezoup Addressed your nits. Please see if my bugfix in virtual_register_count is correct and what action you would like regarding xori and normalize_imm. Thanks!

danielwlz commented on tracer/src/instruction/inline_keccak256/mod.rs:
```diff
@@ -0,0 +1,1010 @@
+
+        // IMPORTANT: XORI can only handle 12-bit immediates in RISC-V
+        // For round constants that don't fit in 12 bits, we need to load them into a register first
+        // TODO: Review this - moodlezoup said that we have 64-bit support for 12-bits, so getting rid of this is a potential optimization.
```
Grabbed the 64imm commit and removed the unneeded code. Thanks!

0xAndoroid replied:
I have the 64bit imm patch in the `feat/rv64` commit b33ad879e89a
